File: go/tools/bash.go
package tools

import (
	"context"
	"fmt"
	"os/exec"
	"time"

	"github.com/steven0lisa/agent-libs/go"
	"github.com/steven0lisa/agent-libs/go/utils"
)

// BashTool executes shell commands with security policy enforcement.
type BashTool struct {
	agentlib.BaseTool
	Whitelist []utils.Pattern
	Blacklist []utils.Pattern
}

// NewBashTool creates a new BashTool with the given whitelist and blacklist.
func NewBashTool(whitelist, blacklist []utils.Pattern) *BashTool {
	return &BashTool{
		Whitelist: whitelist,
		Blacklist: blacklist,
	}
}

// Name returns the tool name.
func (b *BashTool) Name() string { return "bash" }

// Description returns the tool description.
func (b *BashTool) Description() string {
	return "Execute a shell command in the working directory."
}

// InputSchema returns the JSON schema for the tool input.
func (b *BashTool) InputSchema() map[string]any {
	return map[string]any{
		"type": "object",
		"properties": map[string]any{
			"command": map[string]any{
				"type":        "string",
				"description": "The shell command to execute",
			},
			"description": map[string]any{
				"type":        "string",
				"description": "A brief description of what the command does",
			},
			"timeout": map[string]any{
				"type":        "integer",
				"description": "Timeout in milliseconds",
				"default":     120000,
			},
		},
		"required": []string{"command"},
	}
}

// Call executes the bash tool.
func (b *BashTool) Call(ctx context.Context, input map[string]any, toolCtx agentlib.ToolContext) (agentlib.ToolResult, error) {
	command, ok := input["command"].(string)
	if !ok || command == "" {
		return agentlib.Error("command is required"), nil
	}

	// Security check - enforced at execution time, not disclosed in prompt
	allowed, reason := utils.CheckSecurityPolicy(
		command,
		b.Whitelist,
		b.Blacklist,
		true, // default allow
	)
	if !allowed {
		return agentlib.Error(fmt.Sprintf("Command blocked by security policy: %s", reason)), nil
	}

	timeoutMs := agentlib.GetInt(input, "timeout", 120000)

	// Create context with timeout
	execCtx, cancel := context.WithTimeout(ctx, time.Duration(timeoutMs)*time.Millisecond)
	defer cancel()

	// Execute command
	cmd := exec.CommandContext(execCtx, "sh", "-c", command)
	cmd.Dir = toolCtx.WorkDir

	output, err := cmd.CombinedOutput()

	// Check for timeout
	if execCtx.Err() == context.DeadlineExceeded {
		return agentlib.Error(fmt.Sprintf("Command timed out after %dms", timeoutMs)), nil
	}

	isError := err != nil
	return agentlib.ToolResult{
		Content: string(output),
		IsError: isError,
	}, nil
}

File: nodejs/src/tools/bash.ts
/** Bash tool with security policy. */

import { spawn } from 'child_process';
import { ITool, Pattern, ToolContext } from '../config.js';
import { ToolResult, successResult, errorResult } from '../types.js';
import { checkSecurityPolicy } from '../utils/security.js';

export class BashTool implements ITool {
  readonly name = 'bash';
  readonly description = 'Execute a shell command in the working directory.';
  private whitelist: Pattern[];
  private blacklist: Pattern[];

  constructor(whitelist?: Pattern[], blacklist?: Pattern[]) {
    this.whitelist = whitelist || [];
    this.blacklist = blacklist || [];
  }

  readonly inputSchema = {
    type: 'object',
    properties: {
      command: { type: 'string', description: 'The shell command' },
      description: { type: 'string' },
      timeout: { type: 'integer', default: 120000 },
    },
    required: ['command'],
  };

  async call(input: Record<string, unknown>, context: ToolContext): Promise<ToolResult> {
    const command = input.command as string;
    const timeoutMs = (input.timeout as number) ?? 120_000;

    const { allowed, reason } = checkSecurityPolicy(
      command,
      this.whitelist,
      this.blacklist,
      true
    );
    if (!allowed) {
      return errorResult(`Command blocked by security policy: ${reason}`);
    }

    return new Promise<ToolResult>((resolve) => {
      const child = spawn('sh', ['-c', command], {
        cwd: context.workDir,
      });

      let stdout = '';
      let stderr = '';
      let timedOut = false;

      const timer = setTimeout(() => {
        timedOut = true;
        child.kill('SIGKILL');
      }, timeoutMs);

      child.stdout?.on('data', (data: Buffer) => { stdout += data.toString(); });
      child.stderr?.on('data', (data: Buffer) => { stderr += data.toString(); });

      child.on('close', (code: number | null) => {
        clearTimeout(timer);
        if (timedOut) {
          resolve(errorResult(`Command timed out after ${timeoutMs}ms`));
          return;
        }
        const output = stderr ? `${stdout}\n[stderr]\n${stderr}` : stdout;
        resolve(code === 0 ? successResult(output) : errorResult(output));
      });

      child.on('error', (err: Error) => {
        clearTimeout(timer);
        resolve(errorResult(`Failed to execute: ${err.message}`));
      });
    });
  }
}
